### PR TITLE
finished task4

### PR DESCRIPTION
**Pull request by evs062009**

### CODE REVIEW

File: SSA/ElementTasks/src/task4FileParser/Main.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import utilities.IOUtilities;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.stream.Collectors;
+
+public class Main {
+    public static void main(String[] args) {
+        if (args != null && args.length > 1 && args.length < 4) {
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/task4FileParser/Main.java]: @evs062009, what would be the strategy to cover this by tests?

File: SSA/ElementTasks/src/task4FileParser/Main.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        String stringFromFile = Files.lines(Paths.get(file)).collect(
+                Collectors.joining());
+        String stringToFile = stringFromFile.replace(oldString, newString);
+        Files.write(Paths.get(file), stringToFile.getBytes());
+    }
+
+    private static int countStringMatching(String file, String searchString)
+            throws IOException {
+        String stringFromFile = Files.lines(Paths.get(file)).collect(
+                Collectors.joining());
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/task4FileParser/Main.java]: you have the same code in `replaceInFile(...)` method, could it be reused?

File: SSA/ElementTasks/src/task4FileParser/Main.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                    IOUtilities.println(String.format(
+                            "Number of occurrences = %s", count));
+                } else {
+                    replaceInFile(args[0], args[1], args[2]);
+                }
+            } catch (IOException e) {
+                IOUtilities.println(String.format("Invalid file %s.",
+                        e.getMessage()));
+            }
+        } else {
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/task4FileParser/Main.java:16]: could we omit this `else ...` statement?

**Review by nprokhorov — CHANGES_REQUESTED** [review]

File: SSA/ElementTasks/src/task4FileParser/Main.java (hunk 1 above)
Comment (evs062009) [reply SSA/ElementTasks/src/task4FileParser/Main.java]: fixed

File: SSA/ElementTasks/src/task4FileParser/Main.java (hunk 2 above)
Comment (evs062009) [reply SSA/ElementTasks/src/task4FileParser/Main.java]: fixed

File: SSA/ElementTasks/src/task4FileParser/Main.java (hunk 3 above)
Comment (evs062009) [reply SSA/ElementTasks/src/task4FileParser/Main.java:16]: fixed

File: SSA/ElementTasks/src/task4FileParser/services/FileProcessor.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+import utilities.IOUtilities;
+import utilities.StringUtilities;
+
+import java.io.IOException;
+
+public class FileProcessor implements IFileProcessor {
+
+    @Override
+    public void process(String[] args) {
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/task4FileParser/services/FileProcessor.java:11]: What would be the final result if I passed the array of size 4?
and what would happen with elements which index is more than 3?

File: SSA/ElementTasks/src/utilities/Validators.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        message = "is null";
+        return false;
+    }
+
+    public static boolean isStringArrLengthMore(String[] arr, int than) {
+        if (arr.length > than) {
+            clearMessage();
+            return true;
+        }
+        message = String.format("length not more than %s", than);
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/utilities/Validators.java:20]: Seems your method is responsible not only for validation ...

File: SSA/ElementTasks/src/utilities/Validators.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package utilities;
+
+public class Validators {
+    public static String message;
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/utilities/Validators.java:4]: Should the validator have this field?

File: SSA/ElementTasks/src/utilities/StringUtilities.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    }
+
+    public static int countStringMatching(String file, String searchString)
+            throws IOException {
+        String stringFromFile = getStringFromFile(file);
+        return (stringFromFile.length() - stringFromFile.replace(searchString,
+                "").length()) / searchString.length();
+    }
+
+    public static String getStringFromFile (String file) throws IOException {
```
Comment (nprokhorov) [inline SSA/ElementTasks/src/utilities/StringUtilities.java]: What are the method responsibilities? 
What would happen if the file size is greater than 1Gb?

File: SSA/ElementTasks/src/task4FileParser/services/FileProcessor.java (hunk 4 above)
Comment (evs062009) [reply SSA/ElementTasks/src/task4FileParser/services/FileProcessor.java:11]: If you passe the array of size 4, the method will do nothing.
Validation in the main block checks if the array has the correct size. If no, the instruction prints.
https://github.com/evs062009/ElementTasks/blob/438a044e8ba1daf91ab875eef8838ba483924bdb/SSA/ElementTasks/src/task4FileParser/Main.java#L14-L21

File: SSA/ElementTasks/src/utilities/Validators.java (hunk 5 above)
Comment (evs062009) [reply SSA/ElementTasks/src/utilities/Validators.java:20]: It only contains a message, that explains a user the reason if validation is failed. Nothing more.

File: SSA/ElementTasks/src/utilities/Validators.java (hunk 6 above)
Comment (evs062009) [reply SSA/ElementTasks/src/utilities/Validators.java:4]: Yes, to let a user or a programmer know the reason of the failed validation.

File: SSA/ElementTasks/src/utilities/StringUtilities.java (hunk 7 above)
Comment (evs062009) [reply SSA/ElementTasks/src/utilities/StringUtilities.java]: Renamed to convertFileToString(), maybe it is more explicit.
Method responsibilities - convert file contents to a string.
As I understand it depends on JVM settings. We may get OutOfMemoryError.
Added restriction for file size.
https://github.com/evs062009/ElementTasks/blob/35ac584d389601eb1c392bd88124cea6b44fb6c2/SSA/ElementTasks/src/utilities/StringUtilities.java#L27-L33

**Pull request merged**